src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -0,0 +1,536 @@
+                        src2Vector = Sse.Multiply(src1Vector, src2Vector);
+                        Sse.Store(pDstCurrent, src2Vector);
+
+                        pSrc1Current += 4;
+                        pSrc2Current += 4;
+                        pDstCurrent += 4;
+                    }
+
+                    while (pDstCurrent < pEnd)
+                    {
+                        Vector128<float> src1Vector = Sse.LoadScalarVector128(pSrc1Current);
+                        Vector128<float> src2Vector = Sse.LoadScalarVector128(pSrc2Current);
+                        src2Vector = Sse.MultiplyScalar(src1Vector, src2Vector);
+                        Sse.StoreScalar(pDstCurrent, src2Vector);
+
+                        pSrc1Current++;
+                        pSrc2Current++;
+                        pDstCurrent++;
+                    }
+                }
+            }
+        }
+
+        internal static float SumSqU(float[] src, int count)
+        {
+            Vector128<float> result = Sse.SetZeroVector128();
+
+            unsafe
+            {
+                fixed (float* psrc = src)
+                {
+                    float* pSrcCurrent = psrc;
+                    float* pEnd = psrc + count;
+
+                    while (pSrcCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadVector128(pSrcCurrent);
+                        result = Sse.Add(result, Sse.Multiply(srcVector, srcVector));
+
+                        pSrcCurrent += 4;
+                    }
+
+                    if (Sse3.IsSupported)
+                    {
+                        // SSE3 is supported.
+                        result = Sse3.HorizontalAdd(result, result);
+                        result = Sse3.HorizontalAdd(result, result);
+                    }
+                    else
+                    {
+                        // SSE3 is not supported.
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, result));
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, Sse.UnpackLow(result, result)));
+                    }
+
+                    while (pSrcCurrent < pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadScalarVector128(pSrcCurrent);
+                        result = Sse.AddScalar(result, Sse.MultiplyScalar(srcVector, srcVector));
+
+                        pSrcCurrent++;
+                    }
+                }
+            }
+
+            return Sse.ConvertToSingle(result);
+        }
+
+        internal static float SumAbsU(float[] src, int count)
+        {
+            Vector128<float> result = Sse.SetZeroVector128();
+            Vector128<float> mask;
+
+            if (Sse2.IsSupported)
+            {
+                mask = Sse.StaticCast<int, float>(Sse2.SetAllVector128(0x7FFFFFFF));
+            }
+            else
+            {
+                mask = Sse.SetAllVector128(0x7FFFFFFF);
```

Comment:
This isn't right. This will result in `Sse.SetAllVector128((float)(0x7FFFFFFF))`

Reply:
You probably want: `Sse.SetAllVector128(BitConverter.Int32BitsToSingle(0x7FFFFFFF))`

Reply:
Edited thanks!  Is the line above (Line 324) still correct?

Reply:
Any idea?  Thanks!

Reply:
Yes, line 324 is correct since there is an overload in `Sse2` for `public static Vector128<int> SetAllVector128(int value);`.